src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs

```diff
@@ -0,0 +1,77 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+
+namespace System.Text.Json
+{
+    internal sealed class JsonSnakeCaseNamingPolicy : JsonNamingPolicy
+    {
+        public override string ConvertName(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+                return name;
+
+            // Allocates a string builder with the guessed result length,
+            // where 5 is the average word length in English, and
+            // max(2, length / 5) is the number of underscores.
+            StringBuilder builder = new StringBuilder(name.Length + Math.Max(2, name.Length / 5));
+            UnicodeCategory? previousCategory = null;
+
+            for (int currentIndex = 0; currentIndex < name.Length; currentIndex++)
+            {
+                char currentChar = name[currentIndex];
+                if (currentChar == '_')
+                {
+                    builder.Append('_');
+                    previousCategory = null;
+                    continue;
+                }
+
+                UnicodeCategory currentCategory = char.GetUnicodeCategory(currentChar);
+
+                switch (currentCategory)
+                {
+                    case UnicodeCategory.UppercaseLetter:
+                    case UnicodeCategory.TitlecaseLetter:
+                        if (previousCategory == UnicodeCategory.SpaceSeparator ||
+                            previousCategory == UnicodeCategory.LowercaseLetter ||
+                            previousCategory != UnicodeCategory.DecimalDigitNumber &&
+                            currentIndex > 0 &&
+                            currentIndex + 1 < name.Length &&
+                            char.IsLower(name[currentIndex + 1]))
+                        {
+                            builder.Append('_');
+                        }
+
+                        currentChar = char.ToLower(currentChar);
+                        break;
+
+                    case UnicodeCategory.LowercaseLetter:
+                    case UnicodeCategory.DecimalDigitNumber:
+                        if (previousCategory == UnicodeCategory.SpaceSeparator)
+                        {
+                            builder.Append('_');
+                        }
+                        break;
+
+                    case UnicodeCategory.Surrogate:
```

Comment:
Consider adding tests for invalid surrogate characters too (which is possible by doing a substring in between a surrogate pair of characters or crafting one using chars).
So: 
- a string containing a high-surrogate without a low-surrogate following it
- a string containing a low-surrogate first before a high-surrogate

Reply:
Already have one of them. It's 42.

Reply:
The special `42` contains **valid** surrogate pairs (https://www.fileformat.info/info/unicode/char/1d7dc/index.htm).

I am talking about a string that is invalid (i.e. doesn't contain the correct pair of surrogate characters).

For example:
https://github.com/dotnet/corefx/blob/5cee7c97d602f294e27c582d4dab81ec388f1d7b/src/System.Text.Json/tests/JsonDocumentTests.cs#L3545-L3546